src/setup/sound.c

```diff
@@ -129,10 +138,58 @@ static void OpenMusicPackDir(TXT_UNCAST_ARG(widget), TXT_UNCAST_ARG(unused))
     }
 }
 
+#ifdef _WIN32
+static void UpdateMidiDevice(TXT_UNCAST_ARG(widget), TXT_UNCAST_ARG(data))
+{
+    winmm_midi_device = midi_index - 1;
+}
+
+static txt_dropdown_list_t *MidiDeviceSelector(void)
+{
+    txt_dropdown_list_t *result;
+    UINT device_id;
+    UINT num_devices;
+    MMRESULT mmr;
+    MIDIOUTCAPS mcaps;
+    int i;
+
+    num_devices = midiOutGetNumDevs() + 1; // include MIDI_MAPPER
+    if (num_devices > 20)
+    {
+        num_devices = 20;
+    }
+    midi_names = malloc(num_devices * sizeof(char*));
+
+    device_id = -1; // MIDI_MAPPER
+    for (i = 0; i < num_devices; ++i)
+    {
+        mmr = midiOutGetDevCaps(device_id, &mcaps, sizeof(mcaps));
+        if (mmr == MMSYSERR_NOERROR)
+        {
+            midi_names[i] = malloc(sizeof(mcaps.szPname)); // MAXPNAMELEN := 32
+            memcpy(midi_names[i], mcaps.szPname, sizeof(mcaps.szPname));
+        }
+        device_id++;
```

Comment:
This looks odd. Where is `device_id` defined? Since you're incrementing it unconditionally it's effectively the same as 
`device_id = num_devices - 1;`
Did you perhaps mean to only increment inside the `if` ?

Reply:
This bothers me too. I think it should be an array of valid device IDs in case the error check fails. `midi_index` needs some changes too for the same reasons.

Reply:
Done.